### PR TITLE
add support for multilevel susceptibility

### PR DESCRIPTION
**Pull request by oskooi**
Support for multilevel susceptibility is added to model saturable absorbers and lasing thresholds. This involves solving a generalized form of the Maxwell-Bloch equations using second-order accurate finite differences. The parameters are the non-radiative decay and pumping rates of the atomic level transitions as well as the polarization field which have been added to the libctl interface. Tests for a two-level saturable absorber in 1d have verified the threshold behavior and the equal population densities of the two levels at steady state. Note that the fields will blow up if the "sigma" term of the polarization fields or the initial population-level differences are large relative to the temporal resolution. This is not a bug and is related to numerical instabilities arising from strong coupling between the electric and polarization fields. The workaround is to simply increase the spatial resolution or reduce the Courant factor. An example to include in the test suite is forthcoming.

### CODE REVIEW

File: libctl/meep.scm.in [hunk 1]
```diff
@@ -1030,10 +1024,7 @@
 			   (if (null? dt)
 			       (meep-fields-dt-get fields)
 			       (car dt))
-			   (- fcen (/ df 2)) (+ fcen (/ df 2)) maxbands
-                           harminv-spectral-density harminv-Q-thresh
-                           harminv-rel-err-thresh harminv-err-thresh
-                           harminv-rel-amp-thresh harminv-amp-thresh)))
+			   (- fcen (/ df 2)) (+ fcen (/ df 2)) maxbands)))
```
Comment (stevengj) [inline libctl/meep.scm.in]: Why is this change here?

File: libctl/structure.cpp [hunk 2]
```diff
@@ -1300,8 +1297,18 @@ static meep::susceptibility *make_multilevel_sus(const multilevel_atom *d) {
   for (int t = 0; t < d->transitions.num_items; ++t) {
     int i = d->transitions.items[t].from_level - minlev;
     int j = d->transitions.items[t].to_level - minlev;
-    Gamma[i*L+i] += d->transitions.items[t].transition_rate;
-    Gamma[j*L+i] -= d->transitions.items[t].transition_rate;
+    Gamma[i*L+j+1] -= d->transitions.items[t].transition_rate;
```
Comment (stevengj) [inline libctl/structure.cpp]: Why the `+1`?  I thought the `Gamma` array was zero-based?

File: libctl/structure.cpp [hunk 3]
```diff
@@ -1300,8 +1297,18 @@ static meep::susceptibility *make_multilevel_sus(const multilevel_atom *d) {
...
+    Gamma[(i-1)*L+j+1] += d->transitions.items[t].transition_rate;
```
Comment (stevengj) [inline libctl/structure.cpp]: Why isn't `Gamma[i*L+j]` and `Gamma[j*L+i]` correct, i.e. the (i,j) and (j,i) entries of the Γ matrix?

File: libctl/structure.cpp [hunk 4]
```diff
@@ -1300,8 +1297,18 @@ static meep::susceptibility *make_multilevel_sus(const multilevel_atom *d) {
...
+  }
+  
+  // pumping rate matrix Rp
+  meep::realnum *Rp = new meep::realnum[L * L];
+  memset(Rp, 0, sizeof(meep::realnum) * (L*L));
+  for (int t = 0; t < d->transitions.num_items; ++t) {
+    int i = d->transitions.items[t].from_level - minlev;
+    int j = d->transitions.items[t].to_level - minlev;
+    Rp[i*L+i] -= d->transitions.items[t].pumping_rate;
```
Comment (stevengj) [inline libctl/structure.cpp]: `i*L+j`?

File: src/multilevel-atom.cpp [hunk 5]
```diff
@@ -153,13 +159,14 @@ void multilevel_susceptibility::init_internal_data(
...
+  
   d->GammaInv = d->data;
   for (int i = 0; i < L; ++i)
     for (int j = 0; j < L; ++j)
-      d->GammaInv[i*L + j] = (i == j) + Gamma[i*L + j] * dt/2;
+      d->GammaInv[i*L + j] = (i == j) - (Gamma[i*L + j] + Rp[i*L + j]) * dt/2;
+
   if (!invert(d->GammaInv, L)) 
-    abort("multilevel_susceptibility: I + Gamma*dt/2 matrix singular");
+    abort("multilevel_susceptibility: I - (Gamma+R)*dt/2 matrix singular");
```
Comment (stevengj) [inline src/multilevel-atom.cpp]: If R is just added to Γ, why do we need a separate R matrix?  Can't we just modify Γ?

File: libctl/meep.scm.in (hunk 1 above)
Comment (oskooi) [inline libctl/meep.scm.in]: Whoops; I must not have have been working on the most recent version of this file when I made this pull request.

File: libctl/structure.cpp (hunk 2 above)
Comment (oskooi) [inline libctl/structure.cpp]: This is so that the rows of \Gamma add up to zero.

File: src/multilevel-atom.cpp (hunk 5 above)
Comment (oskooi) [inline src/multilevel-atom.cpp]: Having R as a separate matrix makes it easier to add sets of pumping terms between any two levels (there may be more than one pair). This may also simplify things in the future when we make R time dependent.

File: libctl/structure.cpp (hunk 3 above)
Comment (stevengj) [inline libctl/structure.cpp]: Or rather, `i*L+i` and `j*L+i`, the (i,i) and (j,i) entries.

File: libctl/structure.cpp (hunk 2 above)
Comment (stevengj) [inline libctl/structure.cpp]: Don't they still add up to zero if you use the (i,i) and (j,i) entries of Γ?

You need to modify the correct entries as well as having the rows add to zero.

File: libctl/structure.cpp (hunk 4 above)
Comment (stevengj) [inline libctl/structure.cpp]: No, I guess `i*L+i` is correct here.  This should be the same as Γ, above (but see the discussion above about your changes to the Γ indexing).

File: src/multilevel-atom.cpp (hunk 5 above)
Comment (stevengj) [inline src/multilevel-atom.cpp]: I don't understand your comment; regardless of what the interface looks like, couldn't we just combine R and Γ internally?

Comment (oskooi) [inline src/multilevel-atom.cpp]: We can combine R and G internally but I thought it was cleaner to separate them since G should just refer to the non-radiative decay rates.

File: libctl/structure.cpp (hunk 3 above)
Comment (oskooi) [inline libctl/structure.cpp]: Gamma[i*L+j+1] and Gamma[(i-1)*L+j+1] are entries with opposite signs in rows one below the other which thus ensures that the rows sum to zero. The previous version had Gamma[i*L+i] and Gamma[j*L+i] where the rows didn't sum to zero. The entry and sign convention for Gamma is based on eqns. 2-5 of Song's 2011 Optics Express paper.

Comment (oskooi) [inline libctl/structure.cpp]: The "+1" refers to Song's convention of using non-radiative lifetimes with indices where the upper level (first number) appears before the lower level (second number), e.g., \tau_10, \tau_21, etc. This is where these terms populate entries of the Gamma matrix.

File: libctl/structure.cpp (hunk 4 above)
Comment (oskooi) [inline libctl/structure.cpp]: The sign convention of the entries of the pumping-rate matrix Rp are negative for the lower state (negative pump) and positive for the higher one. This is the same convention used in Song's paper and is also consistent with the way this term should be (and is) defined in the interface: from lower state to higher state. This sign convention for the Gamma matrix is flipped since the non-radiative decay is from the higher state to the lower state which is again how it should be and is defined in the interface. The convention that you used in the original formulation of the Gamma matrix had this reversed.

**Comment by oskooi** [comment]
The pumping-rate matrix R has now been absorbed into the Gamma matrix thereby saving storage.

**Comment by oskooi** [comment]
Indexing of the Gamma matrix has been reverted to the original version. All tests pass.
